### PR TITLE
docs: rootfs and kernel image creation guide

### PR DESCRIPTION
**Pull request by dhrgit**
**Description of changes**:
Added a short guide on creating uncompressed kernel and rootfs images, that can be used with Firecracker.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by raduweiss** [comment]
Should we link this from any of the other docs / create an FAQ entry?

**Comment by dhrgit** [comment]
> Should we link this from any of the other docs / create an FAQ entry?

Indeed. I've added a new FAQ entry.

File: docs/rootfs-and-kernel-setup.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+## Creating a rootfs Image
+
+A rootfs image is just a file system image, that hosts at least an init
+system. For instance, our getting started guide uses an EXT4 FS image with
+OpenRC as an init system. Note that, whichever file system you choose to use,
+support for it will have to be compiled into the kernel, so it can be mounted
+at boot time.
+
+To build an EXT4 image:
+1. Prepare a peroperly-sized file. We'll use 50MiB here, but this depends
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
1. Prepare a properly-sized file. We'll use 50 MiB here, but this depends
```

File: docs/rootfs-and-kernel-setup.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+5. Upon a successful build, you can find the uncompressed kernel image under
+   `./vmlinux`.
+
+
+## Creating a rootfs Image
+
+A rootfs image is just a file system image, that hosts at least an init
+system. For instance, our getting started guide uses an EXT4 FS image with
+OpenRC as an init system. Note that, whichever file system you choose to use,
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
`OpenRC` as an init system. Note that, whichever file system you choose to use,
```

(you use `OpenRC` between backtics further in the file, so it's also a matter of consistency)

File: docs/rootfs-and-kernel-setup.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+The final step in the Linux boot process executes `/sbin/init` and expects it
+to never exit. More complex init systems build on top of this, providing
+service configuration files, startup / shutdown scripts for various services,
+and many other features.
+
+For the sake of simplicity, let's set up an Alpine-based rootfs, with `OpenRC`
+as an init system. To that end, we'll use the official Docker image for
+Alpine Linux:
+1. First, let's start the Alpine container, bind-mounting the EXT4 image
+   created earliner, to `/my-rootfs`:
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
   created earlier, to `/my-rootfs`:
```

File: docs/rootfs-and-kernel-setup.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+   ```bash
+   docker run -it --rm -v /tmp/my-rootfs:/my-rootfs alpine
+   ```
+
+2. Then, inside the container, install the OpenRC init system, and some basic
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
2. Then, inside the container, install the `OpenRC` init system, and some basic
```

File: docs/rootfs-and-kernel-setup.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+   ```bash
+   apk add openrc
+   apk add util-linux
+   ```
+
+3. And set up userspace init (still inside the container shell):
+
+   ```bash
+   # Set up a login terminal on the serial console (ttyS0)
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
   # Set up a login terminal on the serial console (ttyS0):
```

File: docs/rootfs-and-kernel-setup.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+   echo ttyS0 > /etc/securetty
+   rc-update add agetty.ttyS0 default
+  
+   # Make sure special file systems are mounted on boot:
+   rc-update add devfs boot
+   rc-update add procfs boot
+  
+   # Then, copy the newly configured system to the rootfs image:
+   for d in bin etc lib root sbin usr; do tar c "/$d" | tar x -C /my-rootfs; done
+   mkdir /my-rootfs/dev
```
Comment (aghecenco) [inline docs/rootfs-and-kernel-setup.md]:
```suggestion
   mkdir /my-rootfs/{dev,proc,run,sys,var}
```

File: docs/rootfs-and-kernel-setup.md (hunk 2 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Removed backticks.

File: docs/rootfs-and-kernel-setup.md (hunk 1 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Fixed.

File: docs/rootfs-and-kernel-setup.md (hunk 3 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Fixed.

File: docs/rootfs-and-kernel-setup.md (hunk 4 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Removed backticks everywhere.

File: docs/rootfs-and-kernel-setup.md (hunk 5 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Fixed.

File: docs/rootfs-and-kernel-setup.md (hunk 6 above)
Comment (dhrgit) [reply docs/rootfs-and-kernel-setup.md]: Done.

**Comment by acatangiu** [comment]
It seems that the unit tests took longer than 240s to run. We should increase that timeout.

```
________________________________ test_unittests ________________________________
integration_tests/build/test_unittests.py:32: in test_unittests
    check=True
/usr/lib/python3.6/subprocess.py:405: in run
    stdout, stderr = process.communicate(input, timeout=timeout)
/usr/lib/python3.6/subprocess.py:835: in communicate
    self.wait()
/usr/lib/python3.6/subprocess.py:1457: in wait
    (pid, sts) = self._try_wait(0)
/usr/lib/python3.6/subprocess.py:1404: in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
E   Failed: Timeout >240.0s
```

**Comment by andreeaflorescu** [comment]
According to our CI we have too many unit tests :scream: 
Probably need to increase the timeout in a different commit and/or open an issue.

**Comment by acatangiu** [comment]
Yes, will do.

**Pull request merged**